src/response-cache.ts

```diff
@@ -8,54 +8,61 @@ interface ResourceCacheContent {
 }
 
 class ResponseCache {
-  cache: Map<string, ResourceCacheContent> = new Map();
+  public readonly cache: Map<string, ResourceCacheContent> = new Map();
 
-  async setCache(response: puppeteer.Response) {
+  public async setCache(response: puppeteer.Response): Promise<void> {
     const url = response.url();
-
     const headers = response.headers();
-    const status =  response.status();
-    const maxAge = this.getMaxAge(headers['cache-control']);
-
-    if (maxAge) {
-      const cacheEntry = this.cache.get(url);
-      if (cacheEntry && cacheEntry.expires < Date.now()) return;
+    const maxAge = ResponseCache.getMaxAge(headers['cache-control']);
 
-      let buffer;
+    if (maxAge && !this.hasNonExpiredCache(url)) {
       try {
-        buffer = await response.buffer();
-      } catch (error) {
-        return;
+        const buffer = await response.buffer();
+        this.cache.set(url, {
+          body: buffer,
+          status: response.status(),
+          headers: headers,
+          expires: Date.now() + (maxAge * 1000)
+        });
+      } catch (e) {
+        // TODO: we should implement a logger.
       }
-
-      this.cache.set(url, {
-        status,
-        headers,
-        body: buffer,
-
-        // todo leak
-        expires: Date.now() + (maxAge * 1000),
-      });
     }
   }
 
-  async request(request: puppeteer.Request) {
+  public async request(request: puppeteer.Request): Promise<boolean> {
```

Comment:
Eğer cache entry varsa ve expire olmuşsa cache içerisinden kaldırmamız gerekiyor. Aksi durumda yukarıdaki mantıkla aynı URL bir daha cache içerisine eklenemiyor. Üstelik memory açısından bu daha sağlıklı. Yine de cache size düşürebilmek için key size düşürmemiz gerekiyor. Bunun için de URL'leri bir şekilde normalize etmeliyiz.